ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureStorageContainerMetricsCollector.java

```diff
@@ -0,0 +1,96 @@
+ * of metrics increases proportionally overwhelming telemetry. This was attempted and failed.
+ *
+ * A daemon will run at regular intervals emitting aggregate metrics in a controlled and predictable manner.
+ * This is a singleton class to avoid multiple collector threads.
+ */
+public class AzureStorageContainerMetricsCollector {
+  private final AzureMetrics metrics;
+  private final ConcurrentHashMap<Long, AzureStorageContainerMetrics> metricMap;
+  private final ScheduledExecutorService executor;
+  private static AzureStorageContainerMetricsCollector instance;
+  private final VerifiableProperties properties;
+
+  public static final Logger logger = LoggerFactory.getLogger(AzureStorageContainerMetricsCollector.class);
+
+  private AzureStorageContainerMetricsCollector(MetricRegistry metrics, VerifiableProperties properties) {
+    metricMap = new ConcurrentHashMap<>();
+    this.metrics = new AzureMetrics(metrics);
+    this.properties = properties;
+    executor = Utils.newScheduler(1, "azure_storage_container_metrics_collector_", true);
+    executor.scheduleWithFixedDelay(getCollector(), 0, 2, TimeUnit.MINUTES);
+    logger.info("Started AzureStorageContainerMetricsCollector");
+  }
+
+  private Runnable getCollector() {
+    return () -> {
+      Long totalDrift = metricMap.values().stream()
+          .map(container -> container.getDrift())
+          .reduce(0L, Long::sum);
+      this.metrics.azureContainerDriftBytesCount.inc(totalDrift);
+    };
+  }
+
+  /**
+   * Thread-safe singleton initializer
+   * @param metrics
+   * @return collector instance
+   */
+  public static synchronized AzureStorageContainerMetricsCollector getInstance(MetricRegistry metrics,
+      VerifiableProperties properties) {
+    if (instance == null) {
+      instance = new AzureStorageContainerMetricsCollector(metrics, properties);
+    }
+    return instance;
+  }
+
+  public void addContainer(Long id) {
+    metricMap.put(id, new AzureStorageContainerMetrics(id));
+  }
+
+  public void removeContainer(Long id) {
+    metricMap.remove(id);
+  }
+
+  /**
+   * Sets the drift of azure-container from ambry-partition.
+   * We use a compare-set to guard against accidental multithreaded errors, although two threads will most likely
+   * not be responsible for a single partition in VCR. A single thread handles all replicas of a partition.
+   * However, we want to avoid any races between reader and writers.
+   * Use min() as bootstrapping replicas can give a wrong picture and indicate a large drift even though the partition
+   * is fully backed up in Azure.
+   * @param id
+   * @param drift
+   */
+  public void setContainerDrift(long id, long drift) {
+    AzureStorageContainerMetrics azureContainerMetrics = metricMap.get(id);
+    Long oldDrift = azureContainerMetrics.getDrift();
+    azureContainerMetrics.setDrift(oldDrift, Math.min(oldDrift, drift));
+  }
```

Comment:
When new partition gets added to the map, the drift value in the AtomicLong would be initialized to 0. But here we are use Math.min to get the smaller value of old drift and new drift, which would always be 0. I don't think this logic would work.

Reply:
The issue is that 0 is a valid drift value, and we are using it to represent an "uninitialized" value. I recommend to use initialize atomic long to -1 and use it as a special case to the values.

Reply:
Good pt ! However, suppose there are 3 replicas with lag as {0, 10, 123456} and assume the last one is a bootstrap replica. I don't want bootstrap replicas skewing this metric as full-auto bootstraps quite a lot. 
I agree with using -1 as an initial value. If we do that, then does it make sense to use `min()` ?

Reply:
Hmm, i see the problem. Once the lag hits 0, it'll stay at 0.